crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -517,6 +526,308 @@ impl<'a> core::iter::FusedIterator for ReserveEntitiesIterator<'a> {}
+    pub fn reserve_entity(&mut self) -> Result<Entity, RemoteReservationError> {
+        if let Some(reserved) = self.current.pop() {
+            return Ok(reserved);
+        }
+
+        if let Some(entities) = self.entities.upgrade() {
+            // SAFETY: `PortableEntities` ensures it is the *only* arc before dropping.
+            return unsafe { Ok((*entities.0).reserve_entity()) };
+        }
+
+        // SAFETY: We just checked that our entities has gone out of scope,
+        // so if users of `PortableEntities` uphold their safety contract, safety is assured.
+        unsafe { self.source.request(RemoteEntityRequest(self.batch_size)) }.map(|fulfilled| {
+            self.current = fulfilled;
+            // SAFETY: the request is non-zero, so there must have been at least one item in it.
+            unsafe { self.current.pop().debug_checked_unwrap() }
+        })
+    }
+
+    /// Returns true if and only if this remote entities is closed.
+    /// For example, this can happen if the source [`Entities`] is cleared or dropped.
+    pub fn is_closed(&self) -> bool {
+        self.source.reserved.is_closed()
+    }
+
+    /// Creates a new [`RemoteEntitiesReserver`] based on this one. This returned reserver will not conflict with this one.
+    ///
+    /// # Safety
+    ///
+    /// The returned reserver must not be used while a `&Entities` is on the thread. This can cause a deadlock.
+    pub unsafe fn new_reserver(&self) -> Self {
+        Self {
+            source: self.source.clone(),
+            current: Vec::new(),
+            entities: self.entities.clone(),
+            batch_size: self.batch_size,
+        }
+    }
+}
+
+impl Drop for RemoteEntitiesReserver {
+    fn drop(&mut self) {
+        if !self.current.is_empty() {
+            // It can't be full, and if it's closed, it doesn't matter.
+            _ = self.source.reserved.push(mem::take(&mut self.current));
+        }
+    }
+}
+
+#[derive(Debug)]
+struct RemoteEntityRequest(NonZero<u32>);
+
+#[derive(Debug)]
+struct RemoteEntitiesSource {
+    generation: AtomicU32,
+    request_more: ConcurrentQueue<RemoteEntityRequest>,
+    reserved: ConcurrentQueue<Vec<Entity>>,
+}
+
+/// An error that occurs when an [`Entity`] can not be reserved remotely.
+/// See also [`RemoteEntitiesReserver`].
+#[derive(thiserror::Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum RemoteReservationError {
+    /// This happens when [`Entities`] are closed, dropped, etc while a [`RemoteEntitiesReserver`] is trying to reserve from it.
+    #[error("A remote entity reserver tried to reserve an entity from a closed `Entities`.")]
+    Closed,
+}
+
+impl RemoteEntitiesSource {
+    /// Requests another chunk of entities.
+    ///
+    /// # Safety
+    ///
+    /// This is a blocking function, so caller must ensure this thread does not have a live reference to `&Entities`,
+    /// preventing the request from being fulfilled on a different thread.
+    unsafe fn request(
```

Comment:
This cannot be a blocking function: it must be async (I think). Bevy can run single threaded (including its asset system), so blocking here would block the asset loading thread and would then block the entire ECS as well.

I think we'll need to use Wakers to listen for when the queue has elements? I'm not fully sure how that works though. If we want a blocking API we can just tell users to use block_on